src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2904,116 +2839,130 @@ private static bool IsReallyAType(TypeSymbol type)
+                        destination = bound;
+                    }
+                    if (!ImplicitConversionExists(source, destination, ref useSiteDiagnostics, conversions.WithNullability(false)))
+                    {
+                        candidates.Remove(candidate);
+                    }
+                    else if (bound.Equals(candidate, TypeCompareKind.IgnoreDynamicAndTupleNames))
+                    {
+                        // SPEC: 4.7 The Dynamic Type
+                        //       Type inference (7.5.2) will prefer dynamic over object if both are candidates.
+                        //
+                        // This rule doesn't have to be implemented explicitly due to special handling of 
+                        // conversions from dynamic in ImplicitConversionExists helper.
+                        // 
+                        if (!TryMergeAndReplaceIfStillCandidate(candidates, candidate, bound, variance, conversions))
+                        {
+                            candidates.Remove(candidate);
+                        }
+                    }
                 }
-                list.Add(type);
             }
+        }
 
-            // Returns true if type was found and merged.
-            internal bool Merge(TypeSymbolWithAnnotations type, ConversionsBase conversions)
+        private static bool TryMergeAndReplaceIfStillCandidate(
+            Dictionary<TypeSymbolWithAnnotations, TypeSymbolWithAnnotations> candidates,
+            TypeSymbolWithAnnotations old,
+            TypeSymbolWithAnnotations @new,
+            VarianceKind variance,
+            ConversionsBase conversions)
+        {
+            // We make an exception when @new is dynamic, for backwards compatibility 
+            if (@new.IsDynamic())
             {
-                // We make an exception when type is dynamic, for backwards compatibility 
-                if (type.IsDynamic())
-                {
-                    return false;
-                }
-
-                List<TypeSymbolWithAnnotations> list;
-                if (!_dictionary.TryGetValue(type, out list))
-                {
-                    return false;
-                }
-
-                bool found = list.Contains(type, _valueComparer);
-                for (int i = 0; i < list.Count; i++)
-                {
-                    list[i] = MergeTupleNames(MergeDynamic(list[i], type, conversions.CorLibrary), type);
-                }
-                return found;
+                return true;
             }
 
-            internal void Remove(TypeSymbolWithAnnotations type)
+            if (candidates.TryGetValue(old, out TypeSymbolWithAnnotations latest))
             {
-                List<TypeSymbolWithAnnotations> list;
-                if (!_dictionary.TryGetValue(type, out list))
+                // Using `variance` seems questionable. What if `latest` was
+                // created by merging two types with a different VarianceKind?
```

Comment:
Is there an issue filed for this?

Reply:
I've updated the comment so it's clear the behavior is expected.

---
In reply to: [224233755](https://github.com/dotnet/roslyn/pull/29727#discussion_r224233755) [](ancestors = 224233755)